src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -50,39 +53,62 @@
+    ColorConsoleAnnotator(String colorMapName, List<AnsiAttributeElement> openTags) {
         this.colorMapName = colorMapName;
-        LOGGER.fine("creating annotator with colorMapName=" + colorMapName);
+        this.openTags = openTags;
+        LOGGER.log(Level.FINE, "creating annotator with colorMapName={0} openTags={2}", new Object[] { colorMapName, openTags });
+    }
+
+    ColorConsoleAnnotator(String colorMapName) {
+        this(colorMapName, Collections.emptyList());
     }
 
     @Override
     public ConsoleAnnotator<Object> annotate(Object context, MarkupText text) {
         String s = text.getText();
-        if (s.indexOf('\u001B') != -1) {
-            AnsiColorMap colorMap = Jenkins.get().getDescriptorByType(AnsiColorBuildWrapper.DescriptorImpl.class).getColorMap(colorMapName);
+        List<AnsiAttributeElement> nextOpenTags = openTags;
+        // TODO: As a performance improvement, we could create a branch where `s.indexOf('\u001B') == -1` but other
+        // conditions are true that surrounds the text in the appropriate tags without going through AnsiHtmlOutputStream.
+        AnsiColorMap colorMap = Jenkins.get().getDescriptorByType(AnsiColorBuildWrapper.DescriptorImpl.class).getColorMap(colorMapName);
+        if (s.indexOf('\u001B') != -1 || !openTags.isEmpty() || colorMap.getDefaultBackground() != null || colorMap.getDefaultForeground() != null) {
             CountingOutputStream outgoing = new CountingOutputStream(new NullOutputStream());
             class EmitterImpl implements AnsiAttributeElement.Emitter {
                 CountingOutputStream incoming;
                 int adjustment;
                 int lastPoint = -1; // multiple HTML tags may be emitted for one control sequence
                 @Override
                 public void emitHtml(String html) {
-                    LOGGER.log(Level.FINEST, "emitting {0} @{1}/{2}", new Object[] { html, incoming.getCount(), s.length() });
-                    text.addMarkup(incoming.getCount(), html);
-                    if (incoming.getCount() != lastPoint) {
-                        lastPoint = incoming.getCount();
-                        int hide = incoming.getCount() - outgoing.getCount() - adjustment;
-                        LOGGER.log(Level.FINEST, "hiding {0} @{1}", new Object[] { hide, outgoing.getCount() + adjustment });
-                        text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<!--", "-->");
-                        adjustment += hide;
+                    int inCount = incoming.getCount();
+                    int outCount = outgoing.getCount() + adjustment;
+                    // All ANSI escapes sequences contain at least 2 bytes on modern platforms, so any HTML emitted
+                    // directly after the first character is received is due to the initialization process of the stream and
+                    // belongs at position 0 (i.e. default background/foreground colors).
```

Comment:
I think I need more caffeine to understand this. :-)

Reply:
Yeah this code was and still is quite confusing to me. Thankfully, the `if` statement is covered by the newly added tests, and if it is removed the tests fail, so we should be able to avoid future regressions.